Failed to apply relocation of type R_AARCH64_TLSLE_ADD_TPREL_HI12 to symbol `_nl_current_LC_COLLATE`

Noticed when running on Arch Linux:
https://github.com/wild-linker/wild/actions/runs/30331825038/job/90190284797?pr=2282

```
---- elf/aarch64/libc-integration/gcc-static ----
Test failed: `libc-integration.c` with linker `wild` config `gcc-static`
  Caused by:
    Linker failed:
wild: error: Failed copying from /usr/lib/gcc/aarch64-linux-gnu/16.1.0/../../../../aarch64-linux-gnu/lib/../lib64/libc.a @ setlocale.o (9730 (38/2)) to output file
  Caused by:
    Failed to apply relocations in section `.text` of /usr/lib/gcc/aarch64-linux-gnu/16.1.0/../../../../aarch64-linux-gnu/lib/../lib64/libc.a @ setlocale.o
    Failed to apply relocation of type R_AARCH64_TLSLE_ADD_TPREL_HI12 to symbol `_nl_current_LC_COLLATE` (11598 local=50) in file #9730 (38/2) (/usr/lib/gcc/aarch64-linux-gnu/16.1.0/../../../../aarch64-linux-gnu/lib/../lib64/libc.a @ setlocale.o) (ABSOLUTE | NON_INTERPOSABLE | DIRECT) at offset 0x9d8
    Relocation -4914064 outside of bounds [0, 16777216)

collect2: error: ld returned 255 exit status

Relink with:
WILD_WRITE_LAYOUT=1 WILD_WRITE_TRACE=1 OUT=/__w/wild/wild/wild/tests/build/elf/aarch64/libc-integration/gcc-static/libc-integration.c.wild /__w/wild/wild/wild/tests/build/elf/aarch64/libc-integration/gcc-static/libc-integration.c.save/run-with cargo run --bin wild --


---- elf/aarch64/libc-integration/gcc-static-pie ----
Test failed: `libc-integration.c` with linker `wild` config `gcc-static-pie`
  Caused by:
    Linker failed:
wild: error: Failed copying from /usr/lib/gcc/aarch64-linux-gnu/16.1.0/../../../../aarch64-linux-gnu/lib/../lib64/libc.a @ setlocale.o (9730 (38/2)) to output file
  Caused by:
    Failed to apply relocations in section `.text` of /usr/lib/gcc/aarch64-linux-gnu/16.1.0/../../../../aarch64-linux-gnu/lib/../lib64/libc.a @ setlocale.o
    Failed to apply relocation of type R_AARCH64_TLSLE_ADD_TPREL_HI12 to symbol `_nl_current_LC_COLLATE` (11572 local=50) in file #9730 (38/2) (/usr/lib/gcc/aarch64-linux-gnu/16.1.0/../../../../aarch64-linux-gnu/lib/../lib64/libc.a @ setlocale.o) (ABSOLUTE | NON_INTERPOSABLE | DIRECT) at offset 0x9d8
    Relocation -768528 outside of bounds [0, 16777216)

collect2: error: ld returned 255 exit status
```

Based on my initial investigation, it's somehow related to fact the symbol has assigned address `0`, and thus `0 - $tp_start` leads to the negative value.
@davidlattimore can you please take a look?

--- a/libwild/src/elf_writer.rs
+++ b/libwild/src/elf_writer.rs
@@ -3411,6 +3411,16 @@ fn apply_relocation<
             .wrapping_add(bias)
             .bitand(mask.got_entry)
             .wrapping_sub(layout.got_base().bitand(mask.got)),
+        RelocationKind::TpOff
+            if layout
+                .symbol_db
+                .is_undefined(layout.symbol_db.definition(local_symbol_id)) =>
+        {
+            // An undefined weak TLS symbol has no offset within the TLS block, so we somewhat
+            // arbitrarily give the 0 offset which at least some other linkers also do and most
+            // importantly is a value guaranteed to fit within the range of any relocation.
+            (addend as u64).wrapping_add(bias)
+        }
         RelocationKind::TpOff => resolution
             .value()
             .wrapping_add(addend as u64)

--- a/wild/tests/sources/elf/undefined-weak-tls/undefined-weak-tls.c
+++ b/wild/tests/sources/elf/undefined-weak-tls/undefined-weak-tls.c
@@ -1,0 +1,23 @@
+//#CompArgs:-ftls-model=local-exec
+//#LinkArgs:-z now --no-gc-sections
+//#Object:runtime.c
+//#Object:ptr_black_box.c
+//#ReferenceLinkers:bfd,lld
+//#SkipArch:ppc64le
+//#ExpectSym:undefined_tls_address
+
+#include "../common/runtime.h"
+
+extern __thread int undefined_tls __attribute__((weak, visibility("hidden")));
+
+// There's not really any way to check any property of this symbol, so we pretty much just make sure
+// that we can link and that the function that references the symbol was included.
+void* undefined_tls_address(void) { return &undefined_tls; }
+
+// Make sure the TLS segment isn't empty.
+__thread int defined_tls;
+
+void _start(void) {
+  runtime_init();
+  exit_syscall(42);
+}

